ztest/ztest.go

```diff
@@ -0,0 +1,219 @@
+//       testdata/
+//         ztest/
+//           test-1.yaml
+//           test-2.yaml
+//           ...
+//       ztest_test.go
+//
+// Name YAML files descriptively since each ztest runs as a subtest
+// named for the file that defines it.
+//
+// ztest_test.go should contain a single Go test that calls Run.
+//
+//     package ztest_test
+//
+//     import (
+//             "testing"
+//
+//             "github.com/brimsec/zq/ztest"
+//     )
+//
+//     func TestZTest(t *testing.T) { ztest.Run(t, "testdata/ztest") }
+//
+// If the ZTEST_ZQ environment variable is unset or empty, Run runs ztests in
+// the current process.  Otherwise, Run run each ztest in a separate process
+// using the zq executable specified by ZTEST_ZQ.
+package ztest
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/brimsec/zq/driver"
+	"github.com/brimsec/zq/emitter"
+	"github.com/brimsec/zq/scanner"
+	"github.com/brimsec/zq/zio/detector"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zql"
+	"github.com/pmezard/go-difflib/difflib"
+	"gopkg.in/yaml.v3"
+)
+
+// Run runs the ztests in the directory named dirname.  For each file f.yaml in
+// the directory, Run calls FromYAMLFile to load a ztest and then runs it in
+// subtest named f.
+func Run(t *testing.T, dirname string) {
+	zq := os.Getenv("ZTEST_ZQ")
+	if zq != "" {
+		if out, err := run(zq, "*", "", "null"); err != nil {
+			if out != "" {
+				out = fmt.Sprintf(" with output %q", out)
+			}
+			t.Fatalf("bad ZTEST_ZQ value %s: %s%s", zq, err, out)
+		}
+	}
+	fileinfos, err := ioutil.ReadDir(dirname)
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, fi := range fileinfos {
+		filename := fi.Name()
+		const dotyaml = ".yaml"
+		if !strings.HasSuffix(filename, dotyaml) {
+			continue
+		}
+		t.Run(strings.TrimSuffix(filename, dotyaml), func(t *testing.T) {
+			t.Parallel()
+			// An absolute path in errors makes the offending file easier to find.
+			filename, err := filepath.Abs(filepath.Join(dirname, filename))
+			if err != nil {
+				t.Fatal(err)
+			}
+			zt, err := FromYAMLFile(filename)
+			if err != nil {
+				t.Fatalf("%s: %s", filename, err)
+			}
+			out, err := run(zq, zt.ZQL, zt.Input, zt.OutputFormat)
+			if err != nil {
+				if out != "" {
+					out = "\noutput:\n" + out
+				}
+				t.Fatalf("%s: %s%s", filename, err, out)
+			}
+			if out != zt.Output {
+				diff, _ := difflib.GetUnifiedDiffString(difflib.UnifiedDiff{
+					A:        difflib.SplitLines(zt.Output),
+					FromFile: "expected",
+					B:        difflib.SplitLines(out),
+					ToFile:   "actual",
+					Context:  5,
+				})
+				t.Fatalf("%s: expected and actual outputs differ:\n%s", filename, diff)
+			}
+		})
+	}
+}
+
+// ZTest defines a ztest.
+type ZTest struct {
+	ZQL          string `yaml:"zql"`
+	Input        string `yaml:"input"`
+	OutputFormat string `yaml:"output-format,omitempty"`
+	Output       string `yaml:"output`
```

Comment:
Missing trailing double-quote for after `"output`.

Reply:
👍